pkg/typesystem/typesystem.go

```diff
@@ -1066,11 +1073,33 @@ func (t *TypeSystem) validateTypeRestrictions(objectType string, relationName st
 				}
 			}
 		}
+
+		if related.Condition != "" {
+			// Validate the conditions referenced by the relations are included in the model.
+			if _, ok := t.conditions[related.Condition]; !ok {
+				return ErrConditionUndefined
+			}
+		}
 	}
 
 	return nil
 }
 
+// validateConditions validates the conditions provided in the model.
+//  1. Each condition should define a well-formed expression,
+//  2. Its parameters should be well-defined
+//  3. The expression and parameters should be compilable.
+func (t *TypeSystem) validateConditions() error {
+	for _, c := range t.conditions {
+		// TODO(jpadilla): cache the compiled condition in the typesystem?
+		_, err := condition.Compile(c)
```

Comment:
We will need to validate the parameters types in here as well (e.g. #2). We only support a handful of parameter types, and if you provide a value other than one of those you should get an error.